[browser] Generate boot config as javascript module

maraf opened the pull request:
This is a first part of https://github.com/dotnet/aspnetcore/issues/59456.
The goal is to still produce `blazor.boot.json` by default, but allow to opt-in for javascript module generation. The schema of the javascript module will change in future PR, for now just moving the JSON to JS.
The PR contains a temporal way (`window["__DOTNET_INTERNAL_BOOT_CONFIG_SRC"]` or `<html data-DOTNET_INTERNAL_BOOT_CONFIG_SRC="">`) to tweak default boot config location for feature opt-in in tests when the PR flows to other repos

maraf committed:
Generate JS module when boot config has .js extension

maraf committed:
Load boot manifest as JS module when URL doesn't contain .json. Temporal way to tweak default boot config location

maraf committed:
Tweak Wasm.Build.Tests to use boot.js

maraf committed:
Tweak Wasm.Build.Tests that create apps from templates to use boot.js

maraf commented:
/azp run runtime

maraf committed:
Fix MiscTests

maraf committed:
Fix RunOutOfAppBundle

maraf committed:
Fix method protection level

maraf committed:
Fix missing head

maraf commented on src/mono/browser/runtime/types/index.ts:
```diff
@@ -244,7 +244,7 @@ export type ResourceList = { [name: string]: string | null | "" };
  * When returned string is not qualified with `./` or absolute URL, it will be resolved against the application base URI.
  */
-export type LoadBootResourceCallback = (type: WebAssemblyBootResourceType, name: string, defaultUri: string, integrity: string, behavior: AssetBehaviors) => string | Promise<Response> | null | undefined;
+export type LoadBootResourceCallback = (type: WebAssemblyBootResourceType, name: string, defaultUri: string, integrity: string, behavior: AssetBehaviors) => string | Promise<Response> | Promise<any> | null | undefined;
```
Great suggestion. Can you add commit with `BootModule` type containing a `config` field of type `MonoConfig`?

pavelsavara replied:
it's not an agent yet, I guess

pavelsavara commented on src/tasks/Microsoft.NET.Sdk.WebAssembly.Pack.Tasks/GenerateWasmBootJson.cs:
```diff
@@ -101,8 +100,7 @@ public override bool Execute()
 
-    // Internal for tests
-    public void WriteBootJson(Stream output, string entryAssemblyName)
+    private void WriteBootJson(string entryAssemblyName)
```
nit: it's not always json anymore, rename ?

pavelsavara commented on src/mono/browser/runtime/loader/config.ts:
```diff
@@ -283,30 +286,53 @@ export function isDebuggingSupported (): boolean {
         loadConfigResponse = await loaderResponse;
+        if (typeof loadConfigResponse.json != "function") {
+            // If the response doesn't contain .json(), consider it an imported module.
+            deep_merge_config(loaderHelpers.config, (loadConfigResponse as any).config);
```
`deep_merge_config` could be the same in all branches, right ? Could we move it down ?

pavelsavara reviewed (commented):
should we also change samples and templates ? 
what about test-main.js ?

pavelsavara commented on src/mono/wasm/Wasm.Build.Tests/BrowserStructures/PublishOptions.cs:
```diff
@@ -26,7 +26,7 @@ public PublishOptions(
         RuntimeVariant RuntimeType                                  = RuntimeVariant.SingleThreaded,
         IDictionary<string, string>? ExtraBuildEnvironmentVariables = null,
-        string BootConfigFileName                                   = "blazor.boot.json",
+        string BootConfigFileName                                   = "boot.js",
```
this should also break expected files in WBT, right ?

maraf commented on src/mono/browser/runtime/loader/config.ts:
```diff
@@ -234,7 +234,10 @@ export async function mono_wasm_load_config (module: DotnetModuleInternal): Prom
-            module.configSrc = "./blazor.boot.json";
+            // Temporal way for tests to opt-in for using boot.js
+            module.configSrc = (globalThis as any)["__DOTNET_INTERNAL_BOOT_CONFIG_SRC"]
+                ?? globalThis.window?.document?.documentElement?.getAttribute("data-dotnet_internal_boot_config_src")
```
It's just yet another temporal alternative, in case I need it anywhere

maraf commented:
> should we also change samples and templates ?
> what about test-main.js ?

Templates are not affected and will benefit from it once it's opt-out.
Library tests are using WasmAppBuilder and thus it's not supported there yet. I'll come with something once it's opt-out.

maraf committed:
Feedback

javiercn commented on src/mono/browser/runtime/loader/config.ts:
```diff
@@ -283,24 +286,41 @@ export function isDebuggingSupported (): boolean {
+            loadedConfigResponse = await fetchBootConfig(appendUniqueQuery(defaultConfigSrc, "manifest"));
+            loadedConfig = await readBootConfigResponse(loadedConfigResponse);
+        } else {
+            loadedConfig = (await import(appendUniqueQuery(defaultConfigSrc, "manifest"))).config;
```
Is this because we are letting people control where `blazor.boot.json` was being loaded from?

javiercn replied:
@maraf is there a need for us to support that moving forward?

What scenario do we think still mandates we keep this around.

I'm saying this because the more we can simplify this process the better. 

Caching and all those things we are handling through the importmap work we did, and I can't really see a reason for this being loaded from elsewhere

pavelsavara replied:
I think this could be in use by non-blazor customers, but I don't have anything specific

maraf replied:
What was the original motivation? Modifying the URLs to point to some CDN or subpath on the server is still valid. 
This callback process all framework assets, not just JS modules

javiercn replied:
Yes, the original motivation was for letting people control the URLs of the deployed app, when we had issues with the anti-virus.

I think it's still fine for things that are "data" like the webcil, but is there a reason why we couldn't rely on relative paths for the module bits? After all, you can already change these urls during the build process modifying the relative paths of the respective assets.

maraf replied:
I don't have a strong opinion. Since blazor users can configure .NET runtime through a callback on `Blazor.start`, they have all the flexibility to define the assets and their URLs in a defferent way than through the loadResource callback, on the other it's a breaking change.

maraf committed:
Update dotnet.boot.js in BuildOptions.cs

maraf committed:
Updatedotnet.boot.js in PublishOptions.cs

pavelsavara commented on src/mono/wasm/testassets/WasmBasicTestApp/App/wwwroot/index.html:
```diff
@@ -7,6 +7,7 @@
   <title>WasmLazyLoading</title>
   <meta charset="UTF-8">
   <meta name="viewport" content="width=device-width, initial-scale=1.0">
+  <script>window["__DOTNET_INTERNAL_BOOT_CONFIG_SRC"] = "boot.js";</script>
```
dotnet.boot.js

maraf replied:
Thank you! 🤦‍♂️

pavelsavara commented on src/tasks/Microsoft.NET.Sdk.WebAssembly.Pack.Tasks/GenerateWasmBootJson.cs:
```diff
@@ -423,7 +421,18 @@ public void WriteBootJson(Stream output, string entryAssemblyName)
         helper.ComputeResourcesHash(result);
-        JsonSerializer.Serialize(output, result, jsonOptions);
+
+        if (Path.GetExtension(OutputPath) == ".js")
```
there is no change that passes `dotnet.boot.js` into this method in the PR, right ?

maraf replied:
Yes, we already had a property that allowed to change boot config file name. It is now set in Wasm.Build.Tests

maraf committed:
Fix remaining places after rename to dotnet.boot.js

pavelsavara committed:
Merge branch 'main' into BootJs

maraf committed:
Now that boot.js starts with dotnet, we need to assert it Wasm.Build.Tests output

maraf committed:
Add missing return statement in method

javiercn commented on src/mono/wasm/Wasm.Build.Tests/ProjectProviderBase.cs:
```diff
@@ -568,17 +568,40 @@ public BootJsonData AssertBootJson(AssertBundleOptions options)
+
+        static BootJsonData LoadConfig(Stream stream)
+        {
+            var serializer = new DataContractJsonSerializer(
+                typeof(BootJsonData),
+                new DataContractJsonSerializerSettings { UseSimpleDictionaryFormat = true });
+
+            var config = (BootJsonData?)serializer.ReadObject(stream);
+            Assert.NotNull(config);
+            return config;
```
Is there a reason why you aren't using S.T.J?

javiercn reviewed (approved):
Looks great!

I have two open questions, but we can address them in a follow-up:
* Getting rid of the old `boot.json` path in the runtime file(s). For 9.0 we should only use `dotnet.boot.js`
* Doing a top-level import for the dotnet.config.js instead of it being updatable. You can already choose the URLs at build time specifying a relative path, and the import can also be relative to the root of the `dotnet.js` file.
* Also, do we think we can get rid of the assets cache now? Or is that a separate item?

maraf commented:
> Getting rid of the old boot.json path in the runtime file(s). For 9.0 we should only use dotnet.boot.js

The goal is to have only `dotnet.boot.js` for .NET 10+.

> Doing a top-level import for the dotnet.config.js instead of it being updatable. You can already choose the URLs at build time specifying a relative path, and the import can also be relative to the root of the dotnet.js file.

We had customers (non blazor) that were interested in not having a boot config as a separate file and configure it manually. I plan to experiment with defining all the modules as static imports to see how it plays with JS bundlers.

> Also, do we think we can get rid of the assets cache now? Or is that a separate item?

Yes, I believe we can remove the cache in a separate task

maraf committed:
Apply default environment name for boot.js

maraf committed:
Skip dotnet.boot.js from checking boot config files starting with dotnet

maraf committed:
Fix double publish in ModuleConfigTests.OverrideBootConfigName

Publish is not incremental and thus config file from first publish
remain in the output after second publish.

Caused by refactoring from https://github.com/dotnet/runtime/pull/109069

maraf committed:
Hotfix computing files to compare stats

maraf committed:
Merge branch 'main' into BootJs
